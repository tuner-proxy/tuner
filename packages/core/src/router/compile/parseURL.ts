const protocol = String.raw`(?<protocol>[^:/?]+:?)`;
const hostname = String.raw`(?<hostname>(\[[^\]]+\]|[^:/?]+))`;
const port = String.raw`(?<port>[^:/?]+)`;
const pathname = String.raw`(?<pathname>\/[^?]*)`;
const search = String.raw`(?<search>\?.*)`;

const host = String.raw`(${hostname}?(:${port})?)`;
const path = String.raw`(?<path>${pathname}?${search}?)`;
const origin = String.raw`(${protocol}?//${host}?)`;

const parser = new RegExp(`^${origin}?${path}?$`);

export interface ParseURLResult {
  protocol?: string;
  hostname?: string;
  port?: string;
  pathname?: string;
  search?: string;
  path?: string;
}

/**
 * Parse URL with fuzzy rules
 *
 * assumptions:
 * - Protocol must have trailing slashes
 * - Pathname must have leading slash
 */
export function parseURL(url: string): ParseURLResult {
  const result = url.match(parser);
  if (result) {
    return result.groups as any;
  }
  throw new Error(`Malformed url '${url}'`);
}
