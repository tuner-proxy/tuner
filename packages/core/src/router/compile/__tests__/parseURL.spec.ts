import { parseURL } from '../parseURL';

describe('parseURL', () => {
  parse('protocol', 'https://');
  parse('multiple protocol', 'https|wss://');

  parse('ip', '//0.0.0.0');
  parse('ip mask', '//[10.0.0.0/16]');
  parse('ip range', '//[10.0.0.0-10.1.2.3]');
  parse('ipv6', '//[2001::1]');
  parse('ipv6 range', '//[2001::abc-2001::1:ffff]');

  parse('hostname', '//example.com');
  parse('dns wildcard', '//*.*.example.com');

  parse('port', '//:80|443');

  parse('pathname', '/example-path/name');
  parse('pathname with route params', '/example-path/:param/::longParam');

  parse('query', '?foo&bar=baz');

  parse(
    'full',
    'https|wss://*.*.example.com:80|443/example-path/name?foo&bar=baz',
  );

  it('throws', () => {
    expect(() => parseURL('malformed-url')).toThrow();
  });
});

function parse(name: string, url: string) {
  it(name, () => {
    expect(parseURL(url)).toMatchSnapshot();
  });
}
