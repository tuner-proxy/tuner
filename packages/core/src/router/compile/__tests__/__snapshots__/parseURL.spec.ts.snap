// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseURL dns wildcard 1`] = `
{
  "hostname": "*.*.example.com",
  "path": undefined,
  "pathname": undefined,
  "port": undefined,
  "protocol": undefined,
  "search": undefined,
}
`;

exports[`parseURL full 1`] = `
{
  "hostname": "*.*.example.com",
  "path": "/example-path/name?foo&bar=baz",
  "pathname": "/example-path/name",
  "port": "80|443",
  "protocol": "https|wss:",
  "search": "?foo&bar=baz",
}
`;

exports[`parseURL hostname 1`] = `
{
  "hostname": "example.com",
  "path": undefined,
  "pathname": undefined,
  "port": undefined,
  "protocol": undefined,
  "search": undefined,
}
`;

exports[`parseURL multiple protocol 1`] = `
{
  "hostname": undefined,
  "path": undefined,
  "pathname": undefined,
  "port": undefined,
  "protocol": "https|wss:",
  "search": undefined,
}
`;

exports[`parseURL pathname 1`] = `
{
  "hostname": undefined,
  "path": "/example-path/name",
  "pathname": "/example-path/name",
  "port": undefined,
  "protocol": undefined,
  "search": undefined,
}
`;

exports[`parseURL pathname with route params 1`] = `
{
  "hostname": undefined,
  "path": "/example-path/:param/::longParam",
  "pathname": "/example-path/:param/::longParam",
  "port": undefined,
  "protocol": undefined,
  "search": undefined,
}
`;

exports[`parseURL port 1`] = `
{
  "hostname": undefined,
  "path": undefined,
  "pathname": undefined,
  "port": "80|443",
  "protocol": undefined,
  "search": undefined,
}
`;

exports[`parseURL protocol 1`] = `
{
  "hostname": undefined,
  "path": undefined,
  "pathname": undefined,
  "port": undefined,
  "protocol": "https:",
  "search": undefined,
}
`;

exports[`parseURL query 1`] = `
{
  "hostname": undefined,
  "path": "?foo&bar=baz",
  "pathname": undefined,
  "port": undefined,
  "protocol": undefined,
  "search": "?foo&bar=baz",
}
`;
